### ISSUE TITLE
Signing task dependency workaround potentially causing problems with test compilation

### ISSUE THREAD

**Issue by JakeWharton** [vanniktech/gradle-maven-publish-plugin#561]
I haven't quite narrowed this completely down, but I think the signing task workaround is causing problems with a test compilation task.

I am seeing:
```
A problem was found with the configuration of task ':molecule-runtime:compileTestKotlinIosSimulatorArm64' (type 'KotlinNativeCompile').
  - Gradle detected a problem with the following location: '/Volumes/dev/cashapp/molecule/molecule-runtime/build/classes/kotlin/iosSimulatorArm64/main/klib'.

    Reason: Task ':molecule-runtime:compileTestKotlinIosSimulatorArm64' uses this output of task ':molecule-runtime:signIosSimulatorArm64Publication' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
```

This looks suspect to me because that location is presumably the output of the main compilation task. I believe this is where the signing workaround dependency is placed, but now the test compilation task is somehow seeing the compilation tasks output (which is its input) as being an output of the signing task.

This is occurring in https://github.com/cashapp/molecule/pull/214 at SHA b9ab64e7073cabfe77bbf209360ee31524870ed1. I'm still working on looking into it but I have to move on for today. Figured I'd preemptively fire the issue, though.

**Comment by JakeWharton** [comment]
Hmm of course it doesn't happen on CI...

**Comment by gabrielittner** [comment]
The workaround only adds a dependency from publish to sign https://github.com/vanniktech/gradle-maven-publish-plugin/blob/main/plugin/src/main/kotlin/com/vanniktech/maven/publish/MavenPublishBaseExtension.kt#L145

> Hmm of course it doesn't happen on CI...

Are you passing the signing secrets to the step that runs the tests on CI? I assume you have them configured globally on your local machine

**Comment by JakeWharton** [comment]
You're right. Locally the signing information is available globally.

**Comment by JakeWharton** [comment]
So my guess here is that the problem is by having signing enabled, the signing task (running after compile, obviously) is placing an `.asc` in the same folder containing the `.klib`. The test compile task has a task input of the folder containing the `.klib` which implicitly gives it a dependency on the main compilation task, but by virtue of writing into that folder Gradle thinks the signing task _should_ be a dependency.

Does that mean this is entirely a Gradle bug? They really shouldn't be placing signing files into the same folder as the output of another task.

**Comment by JakeWharton** [comment]
Or maybe KGP should be using the `.klib` file as a task input rather than the folder? It's really hard to know who's to blame here!

**Comment by gabrielittner** [comment]
I would think this is more on the KGP side because colocating the signature file next to the signed artifact seems very ingrained into Gradle. The Kotlin/JS klib is located in `build/libs` where also all jars are located so maybe KGP could also copy the klib from `build/classes/kotlin` there and use the copy for the publication they are creating to keep it separate.

> Second question: is this something this plugin would be open to configuring working around in the same way it works around the publish/signing dependency problem?
> (The workaround being changing the signing file output folder automatically to a dedicated one)

Yes we can do that here. From a quick look we can have a custom implementation of `SignatureType` that delegates to `ArmoredSignatureType` but has a custom implementation for `fileFor` which returns the `File` to which the signature is written. The `Sign` task can then receive our SignatureType through `setSignatureType`.

**Comment by gabrielittner** [comment]
I've created a quick implementation of the workaround [here](https://github.com/vanniktech/gradle-maven-publish-plugin/compare/signing-workaround) and published `signing-workaround-SNAPSHOT` for testing. It still needs some clean up but it should work and our integration tests succeed.

**Comment by JakeWharton** [comment]
Thanks for looking into it!

I will try to get around to isolating a test project to report to KGP as well, although I can't really estimate a timeframe on that.

**Comment by gabrielittner** [comment]
It took a while to get this released but it's available in 0.25.3
